src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,161 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> lastNode = tail;
+        Node<T> newNode = new Node<>(lastNode, value, null);
+        tail = newNode;
+        if (lastNode == null) {
+            head = newNode;
+        } else {
+            lastNode.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> target = getNode(index);
+            linkBefore(value, target);
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value : list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeAtIndex = head;
+        for (int i = 0; i < index; i++) {
+            nodeAtIndex = nodeAtIndex.next;
+        }
+        return nodeAtIndex.data;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeAtIndex = head;
+        for (int i = 0; i < index; i++) {
+            nodeAtIndex = nodeAtIndex.next;
+        }
+        T oldValue = nodeAtIndex.data;
+        nodeAtIndex.data = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        Node<T> nodeAtIndex = head;
+        for (int i = 0; i < index; i++) {
+            nodeAtIndex = nodeAtIndex.next;
+        }
+        return unlink(nodeAtIndex);
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+        while (current != null) {
+            if (object == null ? current.data == null : object.equals(current.data)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private Node<T> getNode(int index) {
+        checkIndex(index);
+        Node<T> nodeAtIndex = head;
+        for (int i = 0; i < index; i++) {
+            nodeAtIndex = nodeAtIndex.next;
+        }
```

Comment:
![image](https://github.com/mate-academy/jv-linked-list/assets/20326858/6827a183-70b5-48ce-ba00-5b0edefa36ff)

Check the index closer to the tail or to the head before loop.